cmake/ArteryCodegen.cmake

```diff
@@ -0,0 +1,82 @@
+# Artery uses a number of instruments that support generating
+# C++ source and header files from plain text templates, like
+# protobuf or opp_msgc compilers. Often user wants to generate those
+# files before running full build (for static code analysis or linting, for example),
+# so we introduce special artery-codegen target, which aggregates all code generating
+# tasks and runs them together.
+
+set(ARTERY_CODEGEN_TARGET artery-codegen)
+
+if(NOT TARGET ${ARTERY_CODEGEN_TARGET})
+    # Codegen target should be always invoked. Custom commands
+    # are executed as required - it is a responsibility of the codegen
+    # tool provider to ensure sane invocations.
+    add_custom_target(${ARTERY_CODEGEN_TARGET} ALL)
+endif()
+
+macro(artery_mark_as_codegen target)
+    # Marks target as codegen task. Use this on targets that
+    # generate source files (and maybe compile them) only.
+
+    if(NOT TARGET ${target})
+        message(FATAL_ERROR "artery_mark_as_codegen: '${target}' does not exist")
```

Comment:
I suggest to move `add_custom_target(${ARTERY_CODEGEN_TARGET} ...)` here instead of emitting a error message that cannot be reached at all.